### PR TITLE
More control over flow ( interrupt command JIT  + choose commands to ignore/stop on + stop when looping )

### PR DESCRIPTION
**Pull request by eyalk11**
### Background

Now it is quite challenging to control the flow. You would need to assume the next N authorised commands are OK, beforehand. If it goes astray, you have nothing to do, but to wait after the batch. 
This commit make it possible to interrupt just in time and define commands that are more and less relevant with regard to batch.
 
### Changes

1. Added` config.commands_to_stop` and `config.commands_to_ignore`.

`config.commands_to_stop` would stop after the command anyway.  For example, you might want to stop on task completion, anyway. 

 `config.commands_to_ignore` are commands that authorized automatically and don't increase the counter.

2.  Changed spinner class to allow for interruption

  *  `<space>` in the middle of thinking means soft-interrupt complete execution and ask for user input after execution).  
  * `<q>` means abort - exit task immediately and ask for user input.

  Have two implementation for linux and windows.

  Introduced a function `async_task_and_spin`  to wait for either task completion or interrupt. 

3. Stop if already executed a command

  To do so, it keeps all previously executed commands in set. 

### Documentation

In code 

### Test Plan

I just checked the code on my own, saw that both soft and hard interrupts working.
commands_to_stop and to_ignore seems to be working.
There was an error in test_memory_trimmed_from_context that seem unrelated.
```
    memory_found = memory.get_relevant("Important Information", 5)
    assert memory_found[0] == expected_permanent_memory
```
Want to see if it fails on workflow.
### PR Quality Checklist
- [ ] My pull request is atomic and focuses on a single change. (Arguably, you could split this into adding ignore/to stop and adding spinner. But it is part of whole. )
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes 

I didn't add tests because of lack of time. A bit of help is wanted, if you think testing is required for it. 

This is how soft interrupt looks like. It keeps thinking afterwards. 
image

### CODE REVIEW

**Comment by eyalk11** [comment]
There is a tricky thing here. It needs termios and tty for linux and keyboard for windows. as termios failed to install on windows , I couldn't include this in requirments.

**Comment by Pwuts** [comment]
This looks interesting, and like a significant potential UX improvement

**Comment by lc0rp** [comment]
@eyalk11 We're prepping for release v0.4.3. Please resolve conflicts and stand by as we merge.

**Comment by eyalk11** [comment]
OK, I am working on this. it is a bit hard to merge.  (merge_int branch in my fork).

**Comment by eyalk11** [comment]
(There are still some merging issues, will solve)

**Comment by eyalk11** [comment]
I removed the check for reoccurring commands (8139493dd98bab7b474cbbf89b24d6f19275d908) . And added comment in https://github.com/Significant-Gravitas/Auto-GPT/issues/3668 . I will open a new PR for it. 

Spinner for windows nt requires keyboard, currently not in requirements (will fail for linux).

**Comment by eyalk11** [comment]
There is a bug. The bug is that there are leftover executation threads after async_task_and_spin is called because 'pool.shutdown(wait=True)' isn't called. 
Simply doing.
```
loop.call_later(0.1, lambda: pool.shutdown(wait=True)) 
```
in finally won't solve it, because `asyncio.run` will wait until it is called.

That is why, as long as we don't run natively with asyncio, we need another thread to handle it. Otherwise, we could return a future and execute it opportunistically.  My suggested solution is like this: 
https://stackoverflow.com/questions/32059732/send-asyncio-tasks-to-loop-running-in-other-thread

I am open to suggestions. But canceling task without waiting for it to be canceled is problematic.

File: autogpt/spinner.py [hunk 1]
```diff
@@ -13,44 +38,96 @@ def __init__(
...
+
+        self.spinner_thread: RaisingThread = RaisingThread(  # type: ignore
+            target=self.spin
+        )
+        self.spinner_thread.start()  # type: ignore
 
         return self
 
-    def __exit__(self, exc_type, exc_value, exc_traceback) -> None:
+    def __exit__(self, exc_type, exc_value, exc_traceback) -> None:  # type: ignore
```
Comment (Pwuts) [inline autogpt/spinner.py]: why?

File: autogpt/spinner.py [hunk 2]
```diff
@@ -13,44 +38,96 @@ def __init__(
...
             return
+
+        def key_pressed(
+            pressed_key: Optional[str],
+            key_to_check: Optional[str],
+            keyboard_key: Optional[str] = None,
+        ) -> bool:
+            if keyboard_key is None:
+                keyboard_key = key_to_check
+            if (pressed_key and (pressed_key == key_to_check)) or (
```
Comment (Pwuts) [inline autogpt/spinner.py]: nit: `and` takes precedence over `or`, so the extra parentheses here are not necessary

**Review by Pwuts — COMMENTED** [review]
Interrupting continuous execution is currently possible with `Ctrl`+`C`, as implemented by #4230. Is there still more to be done on this front?

The "commands to ignore" and "commands to stop on" are a really good idea, as well as hard/soft interrupts! If you could submit these in separate pull requests, that would be grand (makes it a lot easier to review). :)

File: autogpt/spinner.py (hunk 1 above)
Comment (eyalk11) [reply autogpt/spinner.py]: I was too lazy to check the exact signature. Also not my line, yet prevented merge.

**Comment by eyalk11** [comment]
@Pwuts and @lc0rp I undo refactoring per your request. Consider including this in v0.4.4 . Created branch https://github.com/eyalk11/Auto-GPT/tree/interrupt_commands_with_refactoring that I can PR if you want.

File: autogpt/spinner.py (hunk 1 above)
Comment (eyalk11) [reply autogpt/spinner.py]: fixed `__exit__`

File: autogpt/spinner.py (hunk 2 above)
Comment (eyalk11) [reply autogpt/spinner.py]: just for the user to understand it better..

**Comment by dschonholtz** [comment]
# AI-Maintainer Review for PR - More control over flow (interrupt command JIT + choose commands to ignore/stop on + stop when looping)

## Title and Description :+1:

<summary>The Title and Description are clear, concise, and helpful</summary>
<details>
The title and description of the pull request are clear and concise. They effectively communicate the purpose of the changes, which is to provide more control over the flow in the Auto-GPT project. The description provides a detailed explanation of the changes and their expected behavior.
</details>

## Scope of Changes :+1:

<summary>The changes are narrowly focused</summary>
<details>
The changes in the pull request are narrowly focused on addressing a specific issue related to controlling the flow in the Auto-GPT project. The modifications all contribute to the same goal of providing more control over the flow, and it does not appear that the author is trying to resolve multiple unrelated issues simultaneously.
</details>

## Testing :warning:

<summary>The testing approach could be more comprehensive</summary>
<details>
The author has tested the changes and provided a description of the testing process. However, the testing approach could be more comprehensive. It would be beneficial to include more detailed test cases and potentially automated tests to ensure the new features work as expected in various scenarios.
</details>

## Documentation :warning:

<summary>Some functions, classes, or methods lack docstrings</summary>
<details>
The following functions, classes, or methods lack docstrings:

- `async_task_and_spin` in the `Agent` class
- `start_interaction_loop` in the `Agent` class
- `_resolve_pathlike_command_args` in the `Agent` class
- `spin` in the `Spinner` class
- `update_message` in the `Spinner` class

These should be updated to include docstrings describing their behavior, arguments, and return values.
</details>

## Suggested Changes

- Add docstrings to the functions, classes, or methods mentioned above. The docstrings should describe the behavior, arguments, and return values of each. This will improve code readability and maintainability.
- Consider enhancing the testing approach to include more detailed test cases and potentially automated tests. This will help ensure the new features work as expected in various scenarios and catch any potential issues early.

## Potential Issues

- The `async_task_and_spin` function in the `Agent` class uses asyncio and threading together, which can lead to complex and hard-to-debug issues. Consider refactoring this to use only asyncio or only threading, if possible.
- The `Spinner` class now has an `interruptable` attribute, but it's not clear how this is used if the spinner is not interruptable. Consider adding comments or documentation to clarify this.

**Comment by eyalk11** [comment]
I did restructuring (undoing refactoring) and merged master.

**Comment by eyalk11** [comment]
There is a bug in uploading test results 
```
Run actions/upload-artifact@v3
With the provided path, there will be 47 files uploaded
Starting artifact upload
For more detailed logs during the artifact upload process, enable step-debugging: https://docs.github.com/actions/monitoring-and-troubleshooting-workflows/enabling-debug-logging#enabling-step-debug-logging
Artifact name is valid!
Error: Artifact path is not valid: /DEBUG/20230714_184835_<MagicMock name='mock.ai_name.__getitem__().rstrip()' id='[13](https://github.com/Significant-Gravitas/Auto-GPT/actions/runs/5557322292/jobs/10150954433?pr=3914#step:16:14)9770235010304'>/001/0_full_message_history.json. Contains the following character:  Less than <
          
Invalid characters include:  Double quote ", Colon :, Less than <, Greater than >, Vertical bar |, Asterisk *, Question mark ?, Carriage return \r, Line feed \n
          
The following characters are not allowed in files that are uploaded due to limitations with certain file systems such as NTFS. To maintain file system agnostic behavior, these characters are intentionally not allowed to prevent potential problems with downloads on different file systems.

```
That is why it fails. 

I added test for command_to_stop/ignore.

File: tests/unit/test_ignore_stop_commands.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    clean_input_mock = mocker.patch(
+        "autogpt.agents.agent.clean_input", return_value=config.authorise_key
+    )
+    ai_config = MagicMock()
+    ai_config.ai_name = "Test"
+
+    # with mock.patch("autogpt.workspace.workspace.Workspace", autospec=Workspace):
+    mocker.patch("autogpt.agents.agent.execute_command", side_effect=set_continous)
+    # For some reason config stopped working if it was a decorator
+    with patch.object(
```
Comment (eyalk11) [inline tests/unit/test_ignore_stop_commands.py:69]: As mentioned, for some bloody reason , didn't work  as decorators on the function.

File: autogpt/spinner.py [hunk 4]
```diff
@@ -13,44 +39,101 @@ def __init__(
...
+            pressed_key: Optional[str],
+            key_to_check: Optional[str],
+            keyboard_key: Optional[str] = None,
+        ) -> bool:
+            if keyboard_key is None:
+                keyboard_key = key_to_check
+            if (
+                pressed_key
+                and pressed_key == key_to_check
+                or (os.name == "nt" and keyboard.is_pressed(keyboard_key))
```
Comment (eyalk11) [inline autogpt/spinner.py:82]: not in requirement file

File: requirements.txt [hunk 5]
```diff
@@ -29,6 +29,7 @@ spacy>=3.0.0,<4.0.0
 en-core-web-sm @ https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.5.0/en_core_web_sm-3.5.0-py3-none-any.whl
 prompt_toolkit>=3.0.38
 pydantic
+frozendict
```
Comment (eyalk11) [inline requirements.txt]: keyboard not in req file because isn't supported on mac.

File: autogpt/spinner.py (hunk 4 above)
Comment (eyalk11) [inline autogpt/spinner.py]: will work if out of focus

**Comment by lc0rp** [comment]
@eyalk11 The agent was updated recently, probably the reason for the conflicts.

**Pull request closed without merging**